### ISSUE TITLE
Add a new module for opensrp intergration with external projects

### ISSUE THREAD

**Issue by maimoonak** [OpenSRP/opensrp-server#3]
- Add a new maven module to opensrp server that would facilitate integration of openmrs or any other external system. 
- The module should be able to compile and package as a separate application and should allow to detach itself from main project at any point of time without massive effort.
- The module doesnot need to be fully functional at this point since @mberg would respond about next plan for xlsform extension to incorporate openmrs concept_id. 
- The module basic architecture needs to be part of server application so that further work could be carried out later on and project architecture is complete.

**Comment by maimoonak** [comment]
https://docs.google.com/a/ihsinformatics.com/spreadsheets/d/1L_MeoIo7xP7IckPTKM6ofV023-hPeswid-1M4_iB-6k/edit?usp=sharing
The file has mappings between OpenMRS and OpenSRP fields. It also specifies what data is mandatory when trying to push data to OpenMRS.

https://docs.google.com/a/ihsinformatics.com/spreadsheets/d/1j_UnD6WMtyaDnnlw2cJdiZ8r2_JARFYjfk8zZrK90r4/edit?usp=sharing
The sheet above represents a simple form that registers woman on ANC_Visit. From the sheet openmrs_opensrp we can see the MUST fields (firstname, lastname, gender, birthdate, encounter_date, encounter_location, encounter_provider). We provide openmrs mapping in xlsForm like highlighted columns in sheet. 
- concept shows question in formsubmission mapped with openmrs dictionary item
- patient shows required fields for first time entry for patient registration
- encounter shows fields required related to event/encounter
- parent_id shows concept id for group of concepts under the category.
- in settings there are two attributes showing encounter_type and an attribute showing that form is new

The form would be downloaded as model.xml with attributes embedded inside like few item highlighted in doc here https://docs.google.com/a/ihsinformatics.com/document/d/134Dx0byfpExYxBkmDZ0yVw_mguW5iG8XUHM1o32Obno/edit?usp=sharing

Now we need to figure out that how these attributes would be mapped from model.xml to form_definition.json so that these could be transferred with a formSubmission. The form submission with attributes satisfied by openmrs-connector would be extracted, and pushed to openmrs.

**Comment by maimoonak** [comment]
An abstract breakdown of activities to be conducted for module.
- Mapper from OpenSRP Client and FormSubmission to OpenMRS Patient, Encounter, Obs
- Service Access on 
- Data GET and POST from OpenSRP (from CouchDB) and OpenMRS (REST)
- Event based data push to OpenMRS i.e. Pushas soon as data received on server
- Error Handler and Logger: Log event, event data, and error log incase of failed POSTs. 
- Manual re-POST for errors resolved. As soon as error is resolved allow user/admin to re-post manually.

**Comment by sohelsarder** [comment]
Is that integration module will able to handle submission from other non openSRP client? (i.e mCare case, we need to accept submssions of register clients and their households data collected by mJivita+ application then they what to push that data into openSRP server)

**Comment by ndisha** [comment]
@maimoonak this issue was further broken down into more issues e.g., OpenMRS patient access, OpenMRS encounter access and openmrs core and interface integration..

Is the patient and encounter access done? If yes, can we close it and have the openmrs core and interface integration part referenced in issue #32 ; Error Handler and Logger into another issue and give the milestones accordingly? That way we can know exactly what of this huge component is done and what is remaining.

**Comment by maimoonak** [comment]
After discussion with client/mobile devs in the Sprint, it appeared that our team can not modify anything in ziggy or enketo and trying coordinating with enketo developer would only causing more delay, hence, we should not expect any form submission or form_definition.json specifying the concept mappings. The client devs suggested, that the functionality should be on server side. The idea was more applicable, easier to implement and doable in our timelines. The ticket for the implementation of openmrs attributes parser from form files is here #33 
The final working flow is 
Create an opensrp xls form as specified by doc here https://docs.google.com/document/d/1gcG3FlR1Isdw1YjqOdq4ImfTZFkWa1tIJR0ETnYkz9o/edit
Generate a form.xml, model.xml, form_definition.json and form.json and copy all files in folder opensrp-core/resources/form/formDirName
The formSubmission when submitted to server is parsed with corresponding openmrs attributes and converted to Event and Obs by OpenmrsConnector and then pushed to OpenMRS by opensrp-connector module using the respective services for Patient/Encounter.
Although services were tested and found working fine for very simple test form without repeating groups but actual functionality would be tested and could b marked as working only and only after test real working register. Another ticket would be created for this.
